# 2. Python en acción: tus primeros pasos

Python no solo es un lenguaje de programación poderoso, sino también una herramienta que podemos utilizar de inmediato, sin configuraciones complicadas. En este módulo, daremos nuestros primeros pasos explorando cómo Python puede utilizarse como una calculadora interactiva, permitiéndonos realizar operaciones matemáticas de manera sencilla y directa.

Pero Python no se limita a hacer cálculos. También permite **crear o definir elementos digitales** que podrás usar para organizar la información cuando construimos nuevos programas. Esto es fundamental porque nos da la posibilidad de reutilizar valores, modificarlos, combinarlos en expresiones más complejas y ¡hacerlos interactuar entre sí!. Esta característica, que iremos estudiando en estos módulos, nos permitirá crear programas de la complejidad que se nos antoje (y bueno, es una de las razones por la cual Python es tan popular y útil), pero ten en cuenta que comenzaremos aprendiendo con los elementos más sencillos disponibles tales como números o texto. Estudiaremos cómo funcionan y a diferenciarlos, sentando las bases que nos facilitarán la creación de programas más sofisticados más adelante.


## Metodología del tutorial 

A partir de ahora, cada explicación del tutorial estará usualmente acompañada de casillas de código que deberás inspeccionar (o estudiar) y ejecutar. Tu objetivo siempre debe ser **comprender por qué se produce cada resultado**, de modo que siempre debes tratar de **interpretar los códigos presentados intentando predecir cuál será el resultado de cada ejecución**. Para lograrlo, ¡podrás modificar a tu gusto los ejemplos y verificar sus resultados nuevos! Si tu predicción del resultado de una ejecución resulta correcta, entonces ¡estás aprendiendo el lenguaje! Puedes desarrollar la cantidad de ejemplos que gustes hasta alcanzar el aprendizaje.

También te encontrarás lugares donde se te pedirá, como ejercicio, que desarrolles algún un nuevo ejemplo. **¡No te saltes estos desafíos!** Están pensados para que debas pensar, deducir o construir algo que es clave en tu aprendizaje del lenguaje. Aprender Python es similar a aprender un idioma, deporte, o instrumento musical nuevo: solo leyendo no es suficiente para que puedas hablar el idioma, jugar el deporte, tocar el instrumento o programar en Python. **Es fundamental que pongas en práctica tu conocimiento desde tu mente, tratando de explotar toda tu creatividad**. Te recomiendo que en esta primera etapa te apoyes lo menos posibles en buscadores como Google o IAs (como ChatGPT o Blackbox) que te desarrollen los ejemplos, dado que aquellos serán _simples_ o _poco creativos_, de modo que no te ayudarán a aprender en serio. ¡Desafía tu imaginación! te garantizo que eso marcará la diferencia en tu aprendizaje. 


## 2.1 Python como calculadora

Python puede usarse como una **calculadora interactiva**. Es decir, puedes crear instrucciones que le pidan al _intérprete_ realizar operaciones matemáticas de manera sencilla y directa. Podemos sumar (`+`), restar (`-`), multiplicar (`*`), dividir (`/`) ¡y muchas más!. Prueba estos ejemplos (recuerda que las celdas se ejecutan con `Shift` + `Enter`):

In [ ]:
3+5

8

In [7]:
42.82-14.57

28.25

In [6]:
14*0.6

8.4

In [5]:
1/9

0.1111111111111111

También puede combinar operaciones de forma correcta, respetando el uso de paréntesis y el orden estándar en que las operaciones deben evaluarse a nivel matemático. Por ejemplo: 

In [2]:
3*(6/3 + 1)

9.0

Es decir, Python es capar de evaluar correctamente todos esos ejemplos de memes que circulan por las redes sociales tratando de desafiar a las personas mezclando operaciones matemáticas elementales que deben ser evaluadas siguiendo el orden establecido por las reglas matemáticas. ¡Compruébalo! Aquí está el clásico "desafío" que hasta se ganó [una nota en un portal de noticias]([https://tn.com.ar/tecno/juegos/2024/11/29/cuanto-es-100-10-5-x-2-50-el-desafio-matematico-solo-para-genios/):

In [9]:
100+10/5*2+50

154.0

**Ejercicio**: A continuación tienes algunas celdas de código donde debes crear tres ejemplos _creativos_ que utilicen y/o combinen operaciones matemáticas simples y paréntesis que tú puedas calcular manualmente. ¡Comprueba que el resultado es correcto!

Python también **puede realizar cálculos más complejos**, por ejemplo, usando potencias (`**`), raíces y una infinidad de más cálculos avanzados (hasta matemáticas de niveles universitarios...). Lo mejor de todo es que Python evalúa las expresiones en tiempo real, mostrando el resultado de inmediato, lo que lo convierte en una herramienta ideal para explorar y probar cálculos rápidamente. 

Acá hay algunos ejemplo con potencias. ¿Podrías calcular los resultados a mano antes de ejecutar las celdas? Si no, ¡no importa! Python te dará igualmente el resultado correcto):

In [ ]:
4**2

In [ ]:
2**4

In [ ]:
2**3 + 3**2 

Ahora algo un poco más desafiante para el cálculo manual (desafío digno de meme, pero más interesante que el anterior)...

In [10]:
2*(6/3 + 1)**3

54.0

Si te gustan las matemáticas, te habrás dado cuenta que hasta ahora hemos visto ejemplos que son posibles de resolver a mano (o incluso mentalmente). No obstante, Python nos ofrece la posibilidad de calcular cosas que mentalmente serían un lío... Pensemos en el ejemplo anterior, pero ahora suponiendo que necesitamos que el exponente de la potencia sea un número monstruosamente grande (e elegido 456 por azar). ¿Qué resultado te daría el intérprete de esta operación?

In [11]:
2*(6/3 + 1)**456

7.384517904671098e+217

¡Y violà! Python te devolvió la respueta con la misma rapidez y confianza que con los ejemplos simples. Justamtente, ¡el resultado es número que tiene 218 cifras! El cual lamentablemente no se puede "mostrar" completo por las limitaciones tecnológicas de este tipo de operaciones. En cambio, acá se utiliza la famosa _notación científica_ para representarlo, en la cual se muestran las primeras cifras del número, y se agrupan todas las otras utilizando una potencia de 10. Esto último es indicado con la letra `e` (tanto mayúscula como minúscula) en este lenguaje. Es decir, el número anterior se podría escribir como $7,384 \times 10^{217}$.

En general, se puede utilizar esta `e` para escribir números extremadamente grandes (o pequeños) utilizando esta notación científica. ¡Python no tendrá problemas en operar con ellos tal como lo hace con los otros!, aunque considerando algunas caractarísticas distintas que estudiaremos más adelante. ¿Te quedan dudas de cómo funciona esta forma de escribir números grandes? Observa este ejemplo que simplemente escribe el número _un millón_ utilizándola:

In [14]:
1e6

1000000.0

O algo más entretenido, el número aproximado de estrellas que hay en nuestra galaxia:

In [15]:
1e10

10000000000.0

Para el caso de los números muy chiquitos podemos usar la misma notación. Por ejemplo, consideremos el ancho de un cabello humano, aproximadamente 0.04 milímitros. Para escribir este valor en metros, tenemos que dividirlo por 1000, la cantidad de milímetros que hay en un metro. Esto sería:

In [6]:
0.04/1000

4e-05

¡Pero esto no es todo! Python también incluye algunas operaciones _interesantes_, que no las solemos aplicar en nuestra matemática académica, pero si bastante en la vida cotidiana. Por ejemplo, tiene una forma de calcular el **resto de la división**, un valor muy trascendente cuando realizamos una dividisión manualmente, y que tiene una variedad de aplicaciones dentro de la programación:

In [7]:
6%4

2

## 2.2 Celdas con múltples instrucciones

Hasta ahora hemos utilizado las celdas de código de una forma sencilla: con una única instrucción en Python que nos ha permitido ejecutar operaciones simples y obtener resultados de inmediato. Sin embargo, estos notebooks nos permiten incluir **múltiples instrucciones en una sola celda**, lo que nos da la posibilidad de agrupar cálculos, definir elementos y estructurar mejor nuestro código. ¡Incluso podemos escribir programas completos en una sola celda!

Quizás ya te preguntaste al respecto de esto cuando te encontraste con el código de despedida al final del notebook introductorio. ¿Lo recuerdas? Era algo similar a lo siguiente: 

In [11]:
msg = "Espero que te haya gustado la introducción!"
print(msg)

Espero que te haya gustado la introducción!


La anterior es, justamente, una celda de código que contiene múltiples instrucciones. Acá llegamos a una de las características del funcionamiento de Python que deberás tener siempre en mente: **el intérprete de Python considera que cada línea es una instrucción única a menos que se le indique lo contrario**. De este modo, la celda anterior contiene dos instrucciones distintas: una que almacena un mensaje, y otra que lo muestra como resultado de la ejecución.  

Entonces, cuando ejecutamos una celda que contiene varias instrucciones, Python **lee y ejecuta cada línea de forma independiente, de arriba hacia abajo, en orden**, tal como si estuviera siguiendo una lista de pasos. Cada operación se procesa en secuencia antes de pasar a la siguiente, lo que nos permite realizar cálculos intermedios, almacenar resultados y hacer que diferentes partes del código interactúen entre sí.

#### Comentarios

Cosiderando esta característica, surge la necesidad de poder agregar explicaciones o aclaraciones en texto **entre las instrucciones de una misma celda** para hacer más comprensible el código desarrollado o simplemente mejorar la claridad de lo programado. Para ello, en Python se utiliza el símbolo `#` para indicar que todo el contenido de una instrucción a partir de ese símbolo no debe ser ejecutado, dado que corresponde a un **comentario para humanos** que la máquina debe ignorar en su procesamiento. Por ejemplo, podemos utilizar esto para clarificar nuestro código anterior: 

In [16]:
# Esta línea es un comentario ignorado por el intérprete 

# La siguiente es la primera instrucción de la celda:
msg = "Espero que te haya gustado la introducción!"

# y esta es la segunda instrucción:
print(msg)

Espero que te haya gustado la introducción!


Estos comentarios incluso se pueden utilizar luego de las instrucciones de una misma línea:

In [15]:
356*24   # la cantidad de horas en un año:

8544

Como recomendación general, incluir **comentarios explicativos** es una muy buena práctica al programar. Es decir, es importante tomarse el tiempo de incluir anotaciones en comentarios que expliquen qué hace cada parte del programa. Los comentarios no afectan la ejecución del código, pero son fundamentales para hacerlo más claro y comprensible. Agregar explicaciones breves sobre cálculos, decisiones o estructuras complejas **te ahorrará mucho tiempo cuando necesites revisar tus desarrollos en el futuro**. También facilita que otras personas puedan entender y reutilizar tú código sin necesidad de tener que analizar cada línea en detalle. Un código bien comentado no solo es más fácil de leer, sino que también ayuda a detectar errores y mantener un flujo de trabajo organizado.

#### Resultados de múltiples instrucciones

Considerando lo anterior es importante notar que, por defecto, **el resultado que se muestra al ejecutar una celda con múltiples instrucciones suele ser la salida de la última instrucción escrita**. Revisa el siguiente ejemplo:

In [17]:
# ¿El resuldato de cuál de estas operaciones se muestra? 
4+5
6*8

48

Si tienes dudas, puedes probar _comentando_ la última línea de la celda anterior para verificar que Python está igualmente realizando el cálculo, aunque no esté mostrando su resultado en pantalla. 

Por el contrario, si necesitamos mostrar otros valores intermedios dentro de la misma celda (sin tener que _comentar_ las líneas siguientes), debemos usar la _función_ `print()`. Como ya pudiste ver en el notebook anterior, esta instrucción nos permite mostrar explícitamente cualquier información en cualquier punto del código. Esto es especialmente útil cuando queremos revisar los pasos de un cálculo o analizar el comportamiento del programa. 

La función `print()` muestra en pantalla el elemento que incluyas entre sus paréntesis. Ya hemos visto que para mostrar un texto cualquiera con esta función es necesario encerrarlo con comillas. En el caso de resultados de operaciones, podemos usar igualmente `print()` agregando la operación (o su resultado) dentro de los paréntesis. Por ejemplo:

In [24]:
# Estas tres instrucciones se mostrarán:
print("Este es el resultado de 2+2:")
print(2+2)
print("Listo!")

Este es el resultado de 2+2:
4
Listo!


**EJERCICIO:** A continuación, tienes una celda con múltiples instrucciones incluyendo operaciones matemáticas para que la modifiques a tu gusto: comenta la(s) línea(s) de tu elección, agrega comentarios nuevos o agrega/elimina instrucciones `print()` para que practiques lo estudiado hasta ahora. ¡Ejecútala luego de cada cambio para comprobar qué resultado se muestra al final! ¿Puedes conseguir que se muestre el resultado de la operación $2^3$ sin eliminar los mensajes posteriores?

In [21]:
# recuerda que puedes gregar todos los comentarios que quieras!
2**3

# mostremos un mensaje:
print("Este es un mensaje en medio de la ejecución.")

4+5

5*2

Este es un mensaje en medio de la ejecución.


10

## Variables

En Python podemos definir "variables" que tienen un "nombre" y "contenido":

In [7]:
# Acá tenemos dos definiciones
a = 3
b = 3.1415

In [8]:
#Podemos hacer operaciones entre "variables"
a*b

9.4245

In [9]:
a**b

31.54106995953402

Cada vez que se define una "variable" en Python, en verdad se crean dos cosas:
*   El objeto que contiene la variable.
*   La referencia (nombre) que se usará para acceder al objeto.

Así, cuando hacemos

> a = 3

Python no solo almacena el entero "3" en la memoria, sino que también su nombre "a". Es importante tener en cuenta esto para usar correctamente Python.



Si queremos saber qué es (consultar por el tipo) una variable, en Python podemos utilizar la **función** `type`.

In [10]:
type(3)

int

In [11]:
type(b)

float

In [12]:
type(type)

type

Vamos a utilizar `type` cada vez que no sepamos realmente qué tipo de objeto tengo almacenado en una "variable" (referencia)

In [13]:
# Tip: podemos abreviar algunas operaciones cuando hay variables repetidas
a = 5
a = a + 1
a

6

In [14]:
# Esto se puede escribir como
a = 5
a += 1
a

6

In [15]:
# Python "sabe" (porque el intérprete está en ejecución y no se ha detenido)
# qué variables (referencias) se han creado hasta el momento.
# Por ejemplo, si consultamos por una que no hemos definido:
n

NameError: name 'n' is not defined

In [16]:
# Podemos acceder a la lista de referencias
# definidas con el comando mágico:
%who

a	 b	 


In [17]:
# Como sucedión con el ejemplo de la referencia desconocida,
# Python puede reconocer el tipo de error cuando este sucede. Ejemplo:
zero = 0
a / zero

ZeroDivisionError: division by zero

In [18]:
# Para hacer operaciones matemáticas mas avanzadas,
# utilizamos una librería (o módulo) de Python.
# Todas se cargan con la instrucción import.
# En este caso, usaremos math. :
import math

In [19]:
# veamos el tipo de este objeto:
type(math)

module

In [20]:
# Podemos acceder al contenido de un módulo con el punto:
math.log10(4567.123567)

3.6596427619333136

## Cadenas de caracteres

En Python también puedo definir cadenas de caracteres (conocidas usualmente como _strings_)

In [21]:
s1 = "Hola"
s2 = "Mundo"

In [22]:
# Puedo combinar las variables utilizando las mismas comillas
"Hola"    " "     "Mundo"

'Hola Mundo'

In [23]:
# Incluso puedo realizar operaciones entre las strings:
s1 + s2

'HolaMundo'

In [24]:
s1 + " " + s2 + " cruel"

'Hola Mundo cruel'

In [25]:
#y eso incluso lo podemos guardar en una variable nueva:
s3 = s1 + " " + s2 + " cruel"

In [26]:
s3

'Hola Mundo cruel'